src/components/control/date.jsx

```diff
@@ -0,0 +1,240 @@
+import PropTypes from 'prop-types';
+import React, {
+  forwardRef,
+  useEffect,
+  useLayoutEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from 'react';
+import calendarSvg from '../../svgs/calendar.svg';
+import Calendar from '../calendar/calendar.jsx';
+import IconButton from '../icon-button/icon-button.jsx';
+import Icons from './icons.jsx';
+import useDropdownClose from '../../hooks/use-dropdown-close.js';
+import styles from './date.css';
+import useForwardedRef from '../../hooks/use-forwarded-ref.js';
+
+function toDateStringFormat(date) {
+  return date ? date.toLocaleDateString(undefined, { month: 'short', year: 'numeric', day: 'numeric' }) : '';
+}
+
+function toFullDateFormat(date) {
+  return date ? date.toISOString().slice(0, 10) : undefined;
+}
+
+function fromFullDateFormat(string) {
+  const date = new window.Date(`${string}T00:00:00`);
+  if (date.toDateString() === 'Invalid Date') {
+    return undefined;
+  }
+
+  return date;
+}
+
+const Date = forwardRef(function Date(props, forwardedRef) {
+  const ref = useForwardedRef(forwardedRef);
+  const [active, setActive] = useState(false);
+  const [editing, setEditing] = useState(!!props.validationMessage);
+  const [expanded, setExpanded] = useState(false);
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+  const [value, setValue] = useState(fromFullDateFormat(props.value));
+  const classNames = {
+    container: undefined,
+    input: styles.input,
+    invalidInput: styles['invalid-input'],
+    calendar: styles['calendar-container'],
+    ...props.classNames,
+  };
+  const ids = {
+    input: props.id,
+    label: `${props.id}-label`,
+    tooltip: `${props.id}-tooltip`,
+    validationMessage: `${props.id}Hint`,
+  };
+  const refs = {
+    container: useRef(),
+    input: useRef(),
+  };
+
+  function onBlur(event) {
+    if (refs.container.current.contains(event.relatedTarget)) return;
+
+    // Given the validation spec, any provided validation messages are cleared on blur.
+    // If the native node is still invalid then re-render with the native validation messages.
+    refs.input.current.setCustomValidity('');
+    setValidationMessage(refs.input.current.validationMessage);
+    setActive(false);
+    if (!refs.input.current.validationMessage) setEditing(false);
+  }
+
+  function onInputChange() {
+    setValue(fromFullDateFormat(refs.input.current.value));
+  }
+
+  function onInputClick(event) {
+    if (props.readOnly) return;
+    event.preventDefault();
+    setActive(true);
+    setEditing(true);
+    setExpanded(true);
+  }
+
+  function onInputKeyDown(event) {
+    setActive(true);
+
+    if (props.readOnly) return;
```

Comment:
Should we be able to `setActive` if the input is `readOnly`?

Reply:
🤔 I'm not sure how `active` is used here but a read-only input can still be focusable which I think implies it is active?

Reply:
Fair enough. We can always tweak it later if needed.